unity-renderer/Assets/DCLPlugins/ECS7/Systems/PlayerSystem/ECSPlayerTransformSystem.cs

```diff
@@ -1,78 +1,89 @@
         {
-            public BaseVariable<Transform> avatarTransform;
-            public RendererState rendererState;
-            public Vector3Variable worldOffset;
-            public IReadOnlyList<IParcelScene> loadedScenes;
-            public IECSComponentWriter componentsWriter;
-            public UnityEngine.Vector3 lastAvatarPosition = UnityEngine.Vector3.zero;
-            public Quaternion lastAvatarRotation = Quaternion.identity;
-            public long timeStamp = 0;
+            avatarTransform = DataStore.i.world.avatarTransform;
+            rendererState = CommonScriptableObjects.rendererState;
+            worldOffset = CommonScriptableObjects.worldOffset;
+            loadedScenes = DataStore.i.ecs7.scenes;
+            this.componentsWriter = componentsWriter;
+
+            loadedScenes.OnAdded += LoadedScenesOnOnAdded;
         }
 
-        public static Action CreateSystem(IECSComponentWriter componentsWriter)
+        public void Dispose()
         {
-            var state = new State()
-            {
-                avatarTransform = DataStore.i.world.avatarTransform,
-                rendererState = CommonScriptableObjects.rendererState,
-                worldOffset = CommonScriptableObjects.worldOffset,
-                loadedScenes = DataStore.i.ecs7.scenes,
-                componentsWriter = componentsWriter
-            };
-            return () => Update(state);
+            loadedScenes.OnAdded -= LoadedScenesOnOnAdded;
         }
 
-        private static void Update(State state)
+        public void Update()
         {
-            if (!state.rendererState.Get())
+            if (!rendererState.Get())
                 return;
 
-            Transform avatarT = state.avatarTransform.Get();
+            Transform avatarT = avatarTransform.Get();
 
-            UnityEngine.Vector3 avatarPosition = avatarT.position;
+            Vector3 avatarPosition = avatarT.position;
             Quaternion avatarRotation = avatarT.rotation;
 
-            if (state.lastAvatarPosition == avatarPosition && state.lastAvatarRotation == avatarRotation)
+            if (!newSceneAdded && lastAvatarPosition == avatarPosition && lastAvatarRotation == avatarRotation)
             {
                 return;
             }
 
-            state.lastAvatarPosition = avatarPosition;
-            state.lastAvatarRotation = avatarRotation;
-            UnityEngine.Vector3 worldOffset = state.worldOffset.Get();
+            newSceneAdded = false;
+
+            lastAvatarPosition = avatarPosition;
+            lastAvatarRotation = avatarRotation;
+            Vector3 worldOffset = this.worldOffset.Get();
 
-            var loadedScenes = state.loadedScenes;
-            var componentsWriter = state.componentsWriter;
+            IReadOnlyList<IParcelScene> loadedScenes = this.loadedScenes;
+            var componentsWriter = this.componentsWriter;
 
             IParcelScene scene;
```

Comment:
Move this declaration to the inner scope, you don't use it outside the loop

Reply:
Removed unneeded variables [here](https://github.com/decentraland/unity-renderer/pull/3700/commits/2817506aa3355a9290e0b22d2d080845e50bbae4)